### PR TITLE
feat(preprocessing): R2 — dodot-conflict markers + safety gate

### PR DESCRIPTION
**Pull request by arthur-debert**
## Summary

R2 of the template-magic track from \`docs/proposals/magic.lex\`. Once R3 (\`dodot transform check\`) and R6 (template clean filter) start splicing conflict blocks into template sources to flag ambiguous reverse-merge edits, \`dodot up\` needs to refuse rendering those sources — otherwise the marker lines pass through MiniJinja verbatim and deploy as broken config. R2 lays the constants, the detector, and the pipeline gate so R3/R6 can ship safely.

### Marker shape

\`\`\`
>>>>>> dodot-conflict (template)
host = \"{{ env.DB_HOST }}\"
====== dodot-conflict (deployed)
host = \"production.db.internal\"
<<<<<< dodot-conflict
\`\`\`

6-char angles (not 7) so a file carrying both git's and dodot's markers can be pattern-matched independently. The literal \`dodot-conflict\` token is the load-bearing distinguisher; every config language treats these lines as syntax errors, which is exactly the safety property we want at deploy time.

### What changes

- **\`preprocessing::conflict\` module** with \`MARKER_START\` / \`MARKER_MID\` / \`MARKER_END\` constants, \`contains_unresolved_markers\` / \`find_unresolved_marker_lines\` helpers, and the pipeline-level \`ensure_no_unresolved_markers\` gate.
- **\`DodotError::UnresolvedConflictMarker { source_file, line_numbers }\`** — Display renders the path, the matching line numbers, and a pointer at \`git diff\` for resolution.
- **\`Preprocessor::supports_reverse_merge()\`** trait method (default \`false\`). Templates override to \`true\`. Identity / unarchive stay \`false\` — their sources can't naturally carry the marker token, and reading e.g. a tar archive into a String for the check would be wasteful. Documented contract: \`true\` → MUST populate \`tracked_render\`.
- **Pipeline gate**: \`preprocess_pack\` reads the source bytes and runs the marker scan for every preprocessor with \`supports_reverse_merge() == true\`, before \`expand()\`. On detection: error out, no datastore write, no baseline write.

### Detection details (deliberate choices)

- **Prefix match, not exact equality**: future richer variants like \`>>>>>> dodot-conflict (template, line 12)\` still trip the gate.
- **CRLF-safe**: trailing whitespace stripped before prefix comparison.
- **Line-leading only**: a doc snippet quoting the marker (\`# example: >>>>>> dodot-conflict ...\`) does not trip the gate. Indented mentions also skipped — real reverse-merge output emits markers flush-left.
- **Refuses on any marker**, paired or not. We don't try to validate that markers are balanced; conservative posture beats clever-and-wrong.

## Test plan

- [x] \`cargo test -p dodot-lib --lib\` — **728 tests pass** (no regressions; previously 711, +13 conflict + 4 pipeline gate tests).
- [x] **13 unit tests** in \`preprocessing::conflict\` covering: clean detection, partial blocks, trailing annotation, CRLF, mid-line skip, indented skip, empty content, multi-block, error fields, error message contents, marker-constant directional invariants.
- [x] **4 new pipeline integration tests** cover the gate matrix: markers in template source → blocks expansion **and** no datastore write **and** no baseline write; markers in source for a non-tracking preprocessor → not gated; markers in source for a tracking ScriptedPreprocessor → gated by trait flag (not preprocessor name); clean → blocked → resolved round-trip succeeds again.
- [x] **1 new e2e bats test** for the user-visible CLI behaviour: \`dodot up\` with an unresolved-conflict template surfaces the source path + the \`dodot-conflict\` token in stderr, and no rendered output lands.
- [x] \`cargo clippy --workspace --all-targets -- -D warnings\` — clean.
- [x] \`cargo fmt --all -- --check\` — clean.
- [x] \`bats tests/e2e/bats/test_preprocessing.bats\` — **7/7 pass**.

## Follow-ups

R3 (\`dodot transform check\` reads baselines, computes reverse-diffs via burgertocow, applies the patch with our \`ConflictMarkers\` mapped to the constants from this PR) is the next phase.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: crates/dodot-lib/src/preprocessing/conflict.rs [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+///
+/// On detection, returns a [`DodotError::UnresolvedConflictMarker`]
+/// carrying the source path and the matched line numbers, so the
+/// rendering pipeline reports a clean diagnostic instead of silently
+/// committing a broken config to the deployed location.
+///
+/// `content` is the source bytes interpreted as UTF-8 (lossy). Binary
+/// preprocessor sources (e.g. tar archives) won't naturally contain
+/// the literal marker token; running the check is harmless in that
+/// case.
```
[the remark this answers is not in the document]
Comment (arthur-debert) [reply crates/dodot-lib/src/preprocessing/conflict.rs]: Good catch — fixed both directions:

- **conflict.rs docstring**: corrected to say the API takes valid UTF-8 (it's `&str`), and explicitly directs callers reading from disk to run `String::from_utf8_lossy` first.
- **pipeline.rs gate**: switched from `fs.read_to_string` to `fs.read_file` + `String::from_utf8_lossy`. A reverse-merge-capable preprocessor with a non-UTF-8 source now gets a clean lossy scan instead of a generic UTF-8 decode error.

The marker token is ASCII so lossy decode preserves it. New regression tests in `pipeline.rs::tests::gate_handles_non_utf8_source_via_lossy_decode` and `gate_detects_markers_in_non_utf8_source` pin both directions.

File: crates/dodot-lib/src/preprocessing/pipeline.rs [hunk 2]
```diff
@@ -196,6 +196,21 @@ pub fn preprocess_pack(
...
+        // config. Gated on `supports_reverse_merge` so non-tracking
+        // preprocessors (unarchive, identity) don't pay the read cost
+        // — their sources can't naturally carry the marker token.
+        // See preprocessing-pipeline.lex §6.3.
+        if preprocessor.supports_reverse_merge() {
+            let source_str = fs.read_to_string(&entry.absolute_path)?;
+            crate::preprocessing::conflict::ensure_no_unresolved_markers(
+                &source_str,
+                &entry.absolute_path,
+            )?;
```
[the remark this answers is not in the document]
Comment (arthur-debert) [reply crates/dodot-lib/src/preprocessing/pipeline.rs:220]: Same fix as above: gate now does `fs.read_file` + `String::from_utf8_lossy` so a non-UTF-8 source produces a clean scan (and a real `UnresolvedConflictMarker` if it happens to contain a marker line in valid ASCII), not a filesystem/UTF-8 error. Pinned by two new tests under `gate_handles_non_utf8_source_via_lossy_decode` and `gate_detects_markers_in_non_utf8_source`.

File: crates/dodot-lib/src/error.rs [hunk 3]
```diff
@@ -76,6 +76,12 @@ pub enum DodotError {
     #[error("template variable name \"{name}\" is reserved (dodot and env are built-in namespaces); choose a different name in [preprocessor.template.vars]")]
     TemplateReservedVar { name: String },
 
+    #[error("unresolved dodot-conflict markers in {} at line{} {}\n  resolve the conflict block(s) with `git diff {}` and remove the dodot-conflict marker lines, then re-run.", source_file.display(), if line_numbers.len() == 1 { "" } else { "s" }, line_numbers.iter().map(|n| n.to_string()).collect::<Vec<_>>().join(", "), source_file.display())]
+    UnresolvedConflictMarker {
+        source_file: PathBuf,
+        line_numbers: Vec<usize>,
+    },
```
[the remark this answers is not in the document]
Comment (arthur-debert) [reply crates/dodot-lib/src/error.rs]: Fixed: the hint now renders as `git diff -- '<path>'`. The `--` separator defangs paths starting with a dash; single-quoting handles whitespace and shell metacharacters. Paths containing literal single quotes are not auto-escaped — vanishingly rare in dotfile repos, and the user can adjust the command manually in that edge case (documented in the source).

New tests `error_message_quotes_paths_with_spaces` and `error_message_defangs_leading_dash_paths` pin both protections.
